src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptorMatcher.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.TagHelpers
+{
+    /// <summary>
+    /// Helper methods related to <see cref="TagHelperDescriptor"/> and related classes.
+    /// </summary>
+    public static class TagHelperDescriptorMatcher
+    {
+        /// <summary>
+        /// Determine the full name of the <see cref="Type"/> of the property corresponding to an attribute named
+        /// <paramref name="name"/>.
+        /// </summary>
+        /// <param name="name">Name of the attribute.</param>
+        /// <param name="descriptors">
+        /// Collection of <see cref="TagHelperDescriptor"/>s targeting the containing element in the Razor source.
+        /// </param>
+        /// <returns>
+        /// The full name of the <see cref="Type"/> of the property corresponding to an attribute named
+        /// <paramref name="name"/>. <c>null</c> if the attribute is not bound.
+        /// </returns>
+        /// <remarks>
+        /// This method cannot use Reflection to generally determine the most specific <see cref="Type"/> when
+        /// multiple tag helpers bind the attribute. However when it can determine an attribute is bound to a
+        /// <see cref="string"/> property or dictionary value, it always returns <c>"System.String"</c>. Otherwise it
+        /// returns the full <see cref="Type"/> name for the first binding it finds, if any.
+        /// </remarks>
+        public static string GetPropertyType(
+            [NotNull] string name,
+            [NotNull] IEnumerable<TagHelperDescriptor> descriptors)
+        {
+            // Find first TagHelperAttributeDescriptor matching this name.
+            var firstBoundAttribute = descriptors
+                .SelectMany(descriptor => descriptor.Attributes)
+                .FirstOrDefault(attribute => string.Equals(attribute.Name, name, StringComparison.OrdinalIgnoreCase));
+            if (firstBoundAttribute == null)
+            {
+                return null;
+            }
+
+            if (!IsBoundNonStringAttribute(name, descriptors))
```

Comment:
Actually will unwind some of this (class will be fairly short but I'm not going to remove it) to do the equivalent of `firstBoundAttribute?.IsStringProperty`.  Not adding intentional support for any cases where multiple tag helpers bind the same attribute to properties of different types.  Some scenarios may work but not because we've complicated the code (as I did here) to do it.